### PR TITLE
codemods: migrate viewport related metadata export to viewport export 

### PR DESCRIPTION
**Pull request by huozhi**
This PR adds the codemods transforming certain fields from `metadata` export to `viewport` export, will still preserve rest metadata export properties. We don't cover `generateMetadata` to `generateViewport` as there might be more complex logic inside the function which is hard to apply the codemods rules.

TLDR: Codemods for #57302

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js codemods/viewport-api | Change |
| - | - | - | - |
| buildDuration | 10.1s | 9.9s | N/A |
| buildDurationCached | 5.9s | 6s | ⚠️ +134ms |
| nodeModulesSize | 175 MB | 175 MB | ✓ |
| nextStartRea..uration (ms) | 414ms | 412ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js codemods/viewport-api | Change |
| - | - | - | - |
| 199-HASH.js gzip | 32.3 kB | 32.4 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.2 kB | 53.2 kB | N/A |
| 99.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.5 kB | 45.5 kB | ✓ |
| main-app-HASH.js gzip | 254 B | 251 B | N/A |
| main-HASH.js gzip | 35.3 kB | 35.4 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 45.7 kB | 45.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js codemods/viewport-api | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js codemods/viewport-api | Change |
| - | - | - | - |
| _app-HASH.js gzip | 206 B | 205 B | N/A |
| _error-HASH.js gzip | 182 B | 180 B | N/A |
| amp-HASH.js gzip | 506 B | 505 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.59 kB | 2.59 kB | ✓ |
| edge-ssr-HASH.js gzip | 260 B | 259 B | N/A |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.38 kB | 4.38 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 316 B | 318 B | N/A |
| script-HASH.js gzip | 385 B | 384 B | N/A |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.99 kB | 3.99 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js codemods/viewport-api | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 482 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js codemods/viewport-api | Change |
| - | - | - | - |
| index.html gzip | 528 B | 530 B | N/A |
| link.html gzip | 542 B | 543 B | N/A |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 524 B | 524 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js codemods/viewport-api | Change |
| - | - | - | - |
| edge-ssr.js gzip | 96 kB | 96 kB | N/A |
| page.js gzip | 139 kB | 140 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js codemods/viewport-api | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 622 B | 625 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 25.8 kB | 25.8 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

</details>
Commit: a0c27f4c5bbf58d70770556e692b6a5480d87326

**Comment by ijjk** [comment]
## Failing test suites 
Commit: a0c27f4c5bbf58d70770556e692b6a5480d87326

`pnpm test-dev test/development/middleware-errors/index.test.ts` 
- middleware - development errors > when there is a compilation error from boot > logs the error correctly

<details>
<summary>Expand output</summary>

  ● middleware - development errors › when there is a compilation error from boot › logs the error correctly

    TIMED OUT: success

    undefined

    Error: expect(received).toEqual(expected) // deep equality

    Expected: 2
    Received: 3

      626 |
      627 |   if (hardError) {
    > 628 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      629 |   }
      630 |   return false
      631 | }

      at check (lib/next-test-utils.ts:628:11)
      at Object.<anonymous> (development/middleware-errors/index.test.ts:250:9)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by ztanner — DISMISSED** [review]

File: docs/02-app/01-building-your-application/09-upgrading/01-codemods.mdx [hunk 1]
```diff
@@ -24,6 +24,43 @@ Replacing `<transform>` and `<path>` with appropriate values.
 
 ## Next.js Codemods
 
+### 13.4
```
Comment (huozhi) [inline docs/02-app/01-building-your-application/09-upgrading/01-codemods.mdx:27]:
```suggestion
### 14.0
```

**Review by ztanner — APPROVED** [review]

**Review by shuding — APPROVED** [review]

**Comment by landsman** [comment]
@leerob Please update this blog post with instructions on how to do the update, including running these codemods.
It have to be very hard for people who are not aware of the existence this kind of thing.

I can imagine npm script which will do the update with running codemods for the new version out of the box, what do you think?

**Comment by leerob** [comment]
@landsman https://nextjs.org/docs/app/building-your-application/upgrading/codemods#use-viewport-export

**Pull request merged**